server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -478,6 +478,25 @@ private void executeRequest(
         Rewriteable.rewriteAndFetch(original, searchService.getRewriteContext(timeProvider::absoluteStartMillis), rewriteListener);
     }
 
+    static void adjustSearchType(SearchRequest searchRequest, boolean singleShard) {
+        // optimize search type for cases where there is only one shard group to search on
+        if (singleShard) {
+            // if we only have one group, then we always want Q_T_F, no need for DFS, and no need to do THEN since we hit one shard
+            searchRequest.searchType(QUERY_THEN_FETCH);
+        }
+
+        // if there's only suggest, disable request cache and always use QUERY_THEN_FETCH
+        if (searchRequest.isSuggestOnly()) {
+            searchRequest.requestCache(false);
+            searchRequest.searchType(QUERY_THEN_FETCH);
+        }
+
+        // if there's a kNN search, always use DFS_QUERY_THEN_FETCH
```

Comment:
In a follow-up I'm thinking of handling this in `SearchSourceBuilder#rewrite` instead.